### ISSUE TITLE
[eslint-plugin-react-hooks] New Rule Proposal: `no-unnecessary-setstate `

### ISSUE THREAD

**Issue by karlhorky** [facebook/react#21595]
Hi there!

First of all, again want to extend thanks for your tireless work on creating this amazing library and supporting the ecosystem! Really big up to everyone involved! 💯 🙌

I wanted to see if the `eslint-plugin-react-hooks` project would accept a new rule: `no-unnecessary-setstate`.

I have the feeling that this may be too opinionated for the plugin, but I was sent here by ljharb from `eslint-plugin-react` after proposing it over there: https://github.com/yannickcr/eslint-plugin-react/issues/2997

---

Anyway, here goes:

**Rule proposal:** `no-unnecessary-setstate` - forbid setting state to the value of the existing state variable

I've seen this pattern also a bunch from beginners learning React. And regardless of whether you're a beginner or not, this is probably not what you want to do (even if it's an object / array).

Example of **incorrect** code:

```jsx
function Input (props) {
  const [name, setName] = useState('');	

  return (
    <input
      value={name}
      onChange={(event) => {
        setName(event.currentTarget.value);
        setName(name); // ESLint problem reported
      }}
    />
  );
}
```

Examples of **correct** code:

```jsx
function Input (props) {
  const [name, setName] = useState('');	

  return (
    <input
      value={name}
      onChange={(event) => {
        setName(event.currentTarget.value);
      }}
    />
  );
}
```

```jsx
function Input (props) {
  const [name, setName] = useState('');	

  return (
    <input
      value={name}
      onChange={(event) => {
        const name = event.currentTarget.value; // Shadowing
        setName(name);
      }}
    />
  );
}
```

---

cc @gaearon

**Comment by 1337-ishaan** [comment]
```
  const name = event.currentTarget.value; // Shadowing
        setName(name);
```

Doesn't this cause error? Like variable already declared?

**Comment by karlhorky** [comment]
> Doesn't this cause error? Like variable already declared?

Nope, it's a different scope - this is called "shadowing". I personally don't like this pattern, but it doesn't technically cause problems.

**Comment by 1337-ishaan** [comment]
Maybe because, shadowing makes the code difficult to read.
